domain/port/service/service.go

```diff
@@ -312,3 +261,215 @@ func verifyNoPortRangeConflicts(rangesA, rangesB []network.PortRange) error {
+	ctx domain.AtomicContext, unitUUID unit.UUID, portRanges []network.PortRange,
+) error {
+	colocatedOpened, err := s.st.GetColocatedOpenedPorts(ctx, unitUUID)
+	if err != nil {
+		return errors.Errorf("failed to get opened ports co-located with unit %s: %w", unitUUID, err)
+	}
+	return verifyNoPortRangeConflicts(portRanges, colocatedOpened)
+}
+
+// ensureEndpoints ensures the given endpoints required to open and close a given
+// set of ports are present for the given unit. Returns all endpoints for the unit
+// after ensuring after adding any.
+func (s *Service) addMissingEndpoints(
+	ctx domain.AtomicContext, unitUUID unit.UUID, currentEndpoints []port.Endpoint, openPorts, closePorts network.GroupedPortRanges,
+) ([]port.Endpoint, error) {
+
+	newEndpointNamesSet := set.NewStrings()
+	for endpoint := range openPorts {
+		newEndpointNamesSet.Add(endpoint)
+	}
+	for endpoint := range closePorts {
+		newEndpointNamesSet.Add(endpoint)
+	}
+	for _, endpoint := range currentEndpoints {
+		newEndpointNamesSet.Remove(endpoint.Endpoint)
+	}
+	newEndpointNames := newEndpointNamesSet.SortedValues()
+
+	newEndpoints := []port.Endpoint{}
+	if len(newEndpointNames) > 0 {
+		var err error
+		newEndpoints, err = s.st.AddEndpoints(ctx, unitUUID, newEndpointNames)
+		if err != nil {
+			return nil, errors.Errorf("failed to add endpoints for unit %s: %w", unitUUID, err)
+		}
+	}
+	return newEndpoints, nil
+}
+
+// reconcileWildcard reconciles the open and close port ranges for the wildcard
+// endpoint with the open and close port ranges for all other endpoints, returning
+// a new collection of openPorts and closePorts.
+//
+// There is a special wildcard endpoint "" that represents all endpoints.
+// Any operations applied to the wildcard endpoint will logically applied to all
+// endpoints.
+//
+// That is, if we open a port range on the wildcard endpoint, we will open it as
+// usual. But as a side effect, we ensure that endpoint is not open on any other
+// endpoint.
+//
+// On the other hand, if we close a port range on the wildcard endpoint, we will
+// close it on all other endpoints.
+//
+// Also, if we close a specific endpoint's port range that is open
+// on the wildcard endpoint, we will close it on the wildcard endpoint and open
+// it on all other endpoints except the targeted endpoint.
+//
+// This means:
+//   - Dropping ranges already open on the wildcard endpoint from openPorts for
+//     any endpoint it is present.
+//   - Dropping ranges we are opening on the wildcard endpoint from openPorts for
+//     all other endpoints.
+//   - Adding that range to closePorts for all endpoints (later we we will search
+//     current state to find range uuids to be removed).
+func (s *Service) reconcileWildcard(
+	endpoints []port.Endpoint, openedOnWildcard []network.PortRange, openPortsIn, closePortsIn network.GroupedPortRanges,
+) (network.GroupedPortRanges, network.GroupedPortRanges, error) {
+	openedOnWildcardSet := map[network.PortRange]bool{}
```

Comment:
```suggestion
	openedOnWildcardSet := map[network.PortRange]struct{}
```

Unless the boolean means something, use a struct.